consul/src/main/java/com/linecorp/armeria/internal/consul/ConsulClientBuilder.java

```diff
@@ -0,0 +1,151 @@
+ */
+package com.linecorp.armeria.internal.consul;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.SessionProtocol;
+
+public class ConsulClientBuilder {
+    private static final SessionProtocol DEFAULT_CONSUL_PROTOCOL = SessionProtocol.HTTP;
+    private static final String DEFAULT_CONSUL_ADDRESS = "127.0.0.1";
+    private static final int DEFAULT_CONSUL_PORT = 8500;
+    private static final String DEFAULT_CONSUL_API_VERSION = "v1";
+
+    @Nullable
+    private URI consulUri;
+    private SessionProtocol consulProtocol = DEFAULT_CONSUL_PROTOCOL;
+    private String consulAddress = DEFAULT_CONSUL_ADDRESS;
+    private int consulPort = DEFAULT_CONSUL_PORT;
+    private String consulApiVersion = DEFAULT_CONSUL_API_VERSION;
+    @Nullable
+    private String consulToken;
+    private boolean isUriComponentSet;
+
+    protected ConsulClientBuilder() {
+    }
+
+    /**
+     * Sets the specified Consul's API service URI.
+     * The URI should include the Consul API version at path, like: {@code /v1}
+     * @param consulUri the URI of Consul API service, default: {@code http://127.0.0.1:8500/v1}
+     */
+    public ConsulClientBuilder consulUri(URI consulUri) {
+        checkState(!isUriComponentSet, "consulUri can't comes with other addressing options");
+        requireNonNull(consulUri, "consulUri");
+        final String path = consulUri.getPath();
+        checkArgument(!Strings.isNullOrEmpty(path) && !"/".equals(path),
+                      "consulUri has to contain version path");
+        this.consulUri = consulUri;
+        return this;
+    }
+
+    /**
+     * Sets the specified Consul's API service URI.
+     * The URI should include the Consul API version at path, like: {@code /v1}
+     * @param consulUri the URI of Consul API service, default: {@code http://127.0.0.1:8500/v1}
+     */
+    public ConsulClientBuilder consulUri(String consulUri) {
+        return consulUri(URI.create(requireNonNull(consulUri, "consulUri")));
+    }
+
+    /**
+     * Sets the specified Consul's API service protocol scheme.
+     * @param consulProtocol the protocol scheme of Consul API service, default: {@link SessionProtocol#HTTP}
+     */
+    public ConsulClientBuilder consulProtocol(SessionProtocol consulProtocol) {
+        checkState(consulUri == null, "consulProtocol can't comes with consulUri");
+        requireNonNull(consulProtocol, "consulProtocol");
+        this.consulProtocol = consulProtocol;
+        isUriComponentSet = true;
+        return this;
+    }
+
+    /**
+     * Sets the specified Consul's API service host address.
+     * @param consulAddress the host address of Consul API service, default: {@code 127.0.0.1}
```

Comment:
Sorry, I missed the default values defined in this class.
Could we use {@value ...} expression where possible. e.g. `{@value #DEFAULT_CONSUL_ADDRESS}`

Reply:
Oh, that's great!